### PR TITLE
LIME-2034 Upgrade to Java 25 + Gradle 9.4.1, remove Lombok, and modernise build

### PR DESCRIPTION
**Pull request by smsgds**
### What changed

Build & Tooling
- Upgrade from Java 17 to Java 25 (Corretto) and Gradle 8 to 9.4.1
- Introduce Gradle version catalog (libs.versions.toml) for centralised dependency management
- Move all subproject dependency declarations from inline strings to catalog references
- Consolidate duplicated build boilerplate (JaCoCo, test logging, buildZip) into root build.gradle conventions
- Add AC_TEST_ONLY env var to settings.gradle to allow acceptance-tests to build standalone
- Remove acceptance-tests' own gradle wrapper, settings.gradle, and .sdkmanrc (now uses root)
- Update .sdkmanrc, GitHub Actions workflows to use java-version-file and Corretto distribution
- Bump SAM CLI to 1.157.1, pre-commit hooks, and checkov versions
- Add .hadolint.yaml and hadolint pre-commit hook for Dockerfile linting

Lombok Removal
- Remove Lombok entirely (lombok.config, compileOnly/annotationProcessor deps)
- Replace @Getter, @UtilityClass, @NoArgsConstructor(Lombok) with explicit Java code
- Replace @Data-style classes in lib-dvla with Java records (DvlaPayload, TokenRequestPayload,
  TokenResponse, DriverMatchAPIResponse, DriverMatchErrorResponse, DriverMatchServiceResult,
  RequestHeaderKeys, Errors, etc.)
- Add @Serial and serialVersionUID to all exception classes

Limeade Library Migration
- Replace local Strategy enum, LoggingSupport, SleepHelper, StopWatch, HTTPReply, and
  HTTPConnectionKeepAliveStrategyFactory with equivalents from limeade library
- Replace @ExcludeFromGeneratedCoverageReport with limeade annotations
  (@ExcludeClassFromGeneratedCoverageReport, @ExcludeConstructorFromGeneratedCoverageReport)

Code Quality
- Fix code smells: unused variables, raw types, deprecated API usage (new URL -> URI.create().toURL())
- Replace mutable test field naming (TEST_LIST -> testList) to follow conventions
- Rename validateIntegerRangeData to validateIntegerRangeDataNullIsFail with null handling
- Add new unit tests for DvaCryptographyService, DvaThirdPartyDocumentGateway,
  TokenRequestService, ChangePasswordService, PasswordRenewalHandler, SecretsManagerService

Acceptance Tests
- Consolidate Dockerfile and Dockerfile-traffictest into a single multi-target Dockerfile
  based on Amazon Linux 2023 (gradle:jdk25-corretto), with GPG-verified AWS CLI and Chrome installs
- Consolidate run-tests.sh and run-traffic-tests.sh into a single run-tests.sh using TRAFFIC_TEST env var
- Rework run-tests.sh to run standalone (buildspec bypasses container ENTRYPOINT)
- Remove WebDriverManager dependency; use Selenium's built-in driver management
- Rewrite Driver class: replace ThreadLocal with ConcurrentHashMap, use switch expressions,
  only set Chrome system properties when using Chrome
- Fix race conditions: replace "User clicks on continue" with "User clicks on continue and
  waits for page reload" in scenarios that trigger server-side redirects
- Rename clickAndWait to clickAndWaitForNavigation for clarity
- Align searchForUATUser with standard button click/wait pattern
- Remove redundant "The test is complete and I close the driver" step from all scenarios
  (driver cleanup now handled in @After hook)
- Replace hardcoded 10s waitForVisibility timeouts with configurable MAX_WAIT_SEC
- Increase MAX_WAIT_SEC to 60 for slow redirect chains
- Add rate limiting between tests to prevent overloading smaller F.E. containers
- Modernise build.gradle: convert task/cucumber/smokeTest tasks to tasks.register with JavaExec

CI/CD
- Update all GitHub Actions workflows to read JDK version from .sdkmanrc
- Remove submodules: true from checkout steps
- Run acceptance tests from root with AC_TEST_ONLY=true ./gradlew :acceptance-tests:cucumber
- Add explicit permissions: {} blocks to workflows
- Add .dockerignore at project root

### Why did it change

General maintenance update

### Issue tracking

- [LIME-2034](https://govukverify.atlassian.net/browse/LIME-2034)

### Other considerations

[LIME-2034]: https://govukverify.atlassian.net/browse/LIME-2034?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: lambdas/certexpiryreminder/build.gradle [hunk 1]
```diff
@@ -1,86 +1,143 @@
...
+			libs.bundles.unit.testing
 }
 
-jacocoTestReport {
-	dependsOn test
-	reports {
-		xml.required.set(true)
-	}
-}
+//dependencies {
```
Comment (nlsteers) [inline lambdas/certexpiryreminder/build.gradle:27]: can all this commented out config be removed?

File: .github/workflows/post-merge-package-for-build.yml [hunk 2]
```diff
@@ -14,18 +16,18 @@ jobs:
       ENVIRONMENT: build
     permissions:
       id-token: write
-      contents: read
     steps:
       - name: Check out repository code
         uses: actions/checkout@v6
         with:
           submodules: true
```
Comment (nlsteers) [inline .github/workflows/post-merge-package-for-build.yml:23]: does this repo use submodules?

File: acceptance-tests/Dockerfile [hunk 3]
```diff
@@ -1,25 +1,60 @@
...
+RUN apt-get update \
+    && apt-get install -y --no-install-recommends jq gnupg \
+    && gpg --dearmor < google-linux-signing-key.pub > /usr/share/keyrings/google-chrome.gpg \
+    && echo "deb [arch=amd64 signed-by=/usr/share/keyrings/google-chrome.gpg] http://dl.google.com/linux/chrome/deb/ stable main" > /etc/apt/sources.list.d/google-chrome.list \
+    && apt-get update \
+    && apt-get install -y --no-install-recommends google-chrome-stable \
+    && rm ./google-linux-signing-key.pub \
+    && gpg --import aws-cli.gpg \
+    && gpg --verify awscliv2.zip.sig awscliv2.zip \
+    && unzip awscliv2.zip && ./aws/install \
```
Comment (nlsteers) [inline acceptance-tests/Dockerfile:34]: can you add `-q` flags to unzip? reduces log noise in the output

File: lambdas/passwordRenewal/src/main/java/uk/gov/di/ipv/cri/drivingpermit/event/request/ChangePasswordPayload.java [hunk 4]
```diff
@@ -1,12 +1,35 @@
...
-    private String userName;
-    private String password;
-    private String newPassword;
+public record ChangePasswordPayload(String userName, String password, String newPassword) {
+
+    public static ChangePasswordPayloadBuilder builder() {
+        return new ChangePasswordPayloadBuilder();
+    }
+
+    public static class ChangePasswordPayloadBuilder {
```
Comment (nlsteers) [inline lambdas/passwordRenewal/src/main/java/uk/gov/di/ipv/cri/drivingpermit/event/request/ChangePasswordPayload.java:9]: one to follow up on but could we just update the call sites with direct constructor calls rather than maintaining builder patterns just to keep lombok back compat?

File: lib-dva/src/main/java/uk/gov/di/ipv/cri/drivingpermit/library/dva/configuration/DvaCryptographyServiceConfiguration.java [hunk 5]
```diff
@@ -63,9 +63,7 @@ public class DvaCryptographyServiceConfiguration {
 
     public DvaCryptographyServiceConfiguration(ParameterStoreService parameterStoreService)
             throws CertificateException, NoSuchAlgorithmException {
-        /////////////////
-        //// JWS Map ////
-        /////////////////
+        // JWS Map
```
Comment (nlsteers) [inline lib-dva/src/main/java/uk/gov/di/ipv/cri/drivingpermit/library/dva/configuration/DvaCryptographyServiceConfiguration.java:66]: are these comments really necessary?

File: lib-dvla/src/main/java/uk/gov/di/ipv/cri/drivingpermit/library/dvla/domain/request/RequestHeaderKeys.java [hunk 6]
```diff
@@ -1,11 +1,13 @@
...
-import lombok.experimental.UtilityClass;
-import uk.gov.di.ipv.cri.common.library.annotations.ExcludeFromGeneratedCoverageReport;
+import uk.gov.account.ipv.cri.lime.limeade.annotation.ExcludeClassFromGeneratedCoverageReport;
 
-@UtilityClass
-@ExcludeFromGeneratedCoverageReport
+@ExcludeClassFromGeneratedCoverageReport
 public class RequestHeaderKeys {
+    private RequestHeaderKeys() {
+        /* This utility class should not be instantiated */
```
Comment (nlsteers) [inline lib-dvla/src/main/java/uk/gov/di/ipv/cri/drivingpermit/library/dvla/domain/request/RequestHeaderKeys.java:8]: personal thing but I'd throw an UnsupportedOperationException rather than just an empty constructor with a comment

File: .github/workflows/post-merge-package-for-build.yml [hunk 7]
```diff
@@ -53,35 +55,35 @@ jobs:
...
-          docker build -t $ECR_REGISTRY/$ECR_REPOSITORY_BUILD:$IMAGE_TAG acceptance-tests
-          docker push $ECR_REGISTRY/$ECR_REPOSITORY_BUILD:$IMAGE_TAG
-          cosign sign --yes --tlog-upload=false --key awskms:///${CONTAINER_SIGN_KMS_KEY} $ECR_REGISTRY/$ECR_REPOSITORY_BUILD:$IMAGE_TAG
-          docker tag $ECR_REGISTRY/$ECR_REPOSITORY_BUILD:$IMAGE_TAG $ECR_REGISTRY/$ECR_REPOSITORY_STAGING:$IMAGE_TAG
-          docker push $ECR_REGISTRY/$ECR_REPOSITORY_STAGING:$IMAGE_TAG
-          cosign sign --yes --tlog-upload=false --key awskms:///${CONTAINER_SIGN_KMS_KEY} $ECR_REGISTRY/$ECR_REPOSITORY_STAGING:$IMAGE_TAG
+          docker build -t $ECR_REGISTRY/$ECR_REPOSITORY_TRAFFIC_BUILD:$IMAGE_TAG -f acceptance-tests/Dockerfile --target traffic_test_image .
+          docker push $ECR_REGISTRY/$ECR_REPOSITORY_TRAFFIC_BUILD:$IMAGE_TAG
+          cosign sign --yes --tlog-upload=false --key awskms:///${CONTAINER_SIGN_KMS_KEY} $ECR_REGISTRY/$ECR_REPOSITORY_TRAFFIC_BUILD:$IMAGE_TAG
+          docker tag $ECR_REGISTRY/$ECR_REPOSITORY_TRAFFIC_BUILD:$IMAGE_TAG $ECR_REGISTRY/$ECR_REPOSITORY_TRAFFIC_BUILD:$IMAGE_TAG
```
Comment (nlsteers) [inline .github/workflows/post-merge-package-for-build.yml:68]: this `tag` command just retags the image with the exact same data

**Review by nlsteers — COMMENTED** [review]
looking good so far, a few comments

File: lambdas/certexpiryreminder/build.gradle (hunk 1 above)
Comment (smsgds) [reply lambdas/certexpiryreminder/build.gradle]: TY forgot to remove this reference

File: .github/workflows/post-merge-package-for-build.yml (hunk 2 above)
Comment (smsgds) [reply .github/workflows/post-merge-package-for-build.yml]: Removed, will do the same in Passport and Fraud - last submodules where common lib and common lambdas

File: lambdas/passwordRenewal/src/main/java/uk/gov/di/ipv/cri/drivingpermit/event/request/ChangePasswordPayload.java (hunk 4 above)
Comment (smsgds) [reply lambdas/passwordRenewal/src/main/java/uk/gov/di/ipv/cri/drivingpermit/event/request/ChangePasswordPayload.java:9]: I like the builder pattern it keeps the code very clean, but the aim here was simply remove lombok.

File: lib-dva/src/main/java/uk/gov/di/ipv/cri/drivingpermit/library/dva/configuration/DvaCryptographyServiceConfiguration.java (hunk 5 above)
Comment (smsgds) [reply lib-dva/src/main/java/uk/gov/di/ipv/cri/drivingpermit/library/dva/configuration/DvaCryptographyServiceConfiguration.java:66]: Leaving for now, this API will be removed and they may still help.

File: .github/workflows/post-merge-package-for-build.yml (hunk 7 above)
Comment (smsgds) [reply .github/workflows/post-merge-package-for-build.yml]: I suspect this line was originally copied from the AC test image tagging for staging

File: .github/workflows/post-merge-deploy-to-dev.yml [hunk 8]
```diff
@@ -61,13 +60,20 @@ jobs:
...
+      - name: Build, tag, and push testing images to Amazon ECR
+        env:
+          CONTAINER_SIGN_KMS_KEY_DEV: ${{ secrets.CONTAINER_SIGN_KMS_KEY_DEV }}
+          ECR_REGISTRY: ${{ steps.login-ecr.outputs.registry }}
+          ECR_REPOSITORY_DEV: ${{ secrets.ECR_REPOSITORY_DEV }}
+          IMAGE_TAG: latest
+        run: |
+          docker build -t $ECR_REGISTRY/$ECR_REPOSITORY_DEV:$IMAGE_TAG -f acceptance-tests/Dockerfile --target ac_test_image .
+          docker push $ECR_REGISTRY/$ECR_REPOSITORY_DEV:$IMAGE_TAG
+          cosign sign --yes --tlog-upload=false --key awskms:///${CONTAINER_SIGN_KMS_KEY_DEV} $ECR_REGISTRY/$ECR_REPOSITORY_DEV:$IMAGE_TAG
```
Comment (nlsteers) [inline .github/workflows/post-merge-deploy-to-dev.yml]: this entire job can be replaced with this

```yml
  test-image-build-and-push:
    name: Build and push test image
    runs-on: ubuntu-latest
    steps:
      - uses: govuk-one-login/devplatform-upload-action-ecr@f82c8774a6390f200b15112b3cbb9faf3442173c
        with:
          role-to-assume-arn: ${{ secrets.role_arn }}
          build-and-push-image-only: true
          artifact-bucket-name: ''
          ecr-repo-name: ${{ secrets.ecr_repository_tests }}
          dockerfile: my-awesome.Dockerfile
          push-latest-tag: true
          container-sign-kms-key-arn: ${{ secrets.container_sign_kms_key }}
```

File: .github/workflows/post-merge-deploy-acm-to-dev.yml [hunk 9]
```diff
@@ -23,14 +23,12 @@ jobs:
...
         uses: actions/checkout@v6
-        with:
-          submodules: true
 
       - name: Setup SAM
         uses: aws-actions/setup-sam@d78e1a4a9656d3b223e59b80676a797f20093133 # v2
         with:
           use-installer: true
-          version: "1.134.0"
+          version: "1.157.1"
```
Comment (nlsteers) [inline .github/workflows/post-merge-deploy-acm-to-dev.yml:31]: SAM is already available on `ubuntu-latest`

File: .github/workflows/post-merge-deploy-to-dev.yml [hunk 10]
```diff
@@ -61,13 +60,20 @@ jobs:
           ECR_REPOSITORY_TRAFFIC_DEV: ${{ secrets.ECR_REPOSITORY_TRAFFIC_DEV  }}
           IMAGE_TAG: latest
         run: |
-          docker build -t $ECR_REGISTRY/$ECR_REPOSITORY_TRAFFIC_DEV:$IMAGE_TAG -f acceptance-tests/Dockerfile-traffictest acceptance-tests
+          docker build -t $ECR_REGISTRY/$ECR_REPOSITORY_TRAFFIC_DEV:$IMAGE_TAG -f acceptance-tests/Dockerfile --target ac_test_image .
```
Comment (nlsteers) [inline .github/workflows/post-merge-deploy-to-dev.yml:63]: shouldn't the target be `traffic_test_image` here?

File: acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/utilities/Driver.java [hunk 11]
```diff
@@ -72,34 +69,46 @@ public static WebDriver get() {
...
     }
 
     public static void closeDriver() {
-        if (driverPool.get() != null) {
-            driverPool.get().close();
-            driverPool.get().quit();
+        WebDriver driver = driverPool.get();
+        if (driver != null) {
             driverPool.remove();
+            closeAllDrivers();
```
Comment (nlsteers) [inline acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/utilities/Driver.java:100]: any test calling `closeDriver()` will terminate drivers belonging to any other thread, is this what was intended?

File: acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/utilities/Driver.java [hunk 12]
```diff
@@ -72,34 +69,46 @@ public static WebDriver get() {
...
+        }
+    }
+
+    private static void closeAllDrivers() {
+        for (WebDriver driver : activeDrivers) {
+            activeDrivers.remove(driver);
```
Comment (nlsteers) [inline acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/utilities/Driver.java:106]: I think this will try and close the already closed driver on L99

File: acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/utilities/Driver.java [hunk 13]
```diff
@@ -72,34 +69,46 @@ public static WebDriver get() {
...
-            driverPool.get().quit();
+        WebDriver driver = driverPool.get();
+        if (driver != null) {
             driverPool.remove();
+            closeAllDrivers();
+        }
+    }
+
+    private static void closeAllDrivers() {
+        for (WebDriver driver : activeDrivers) {
```
Comment (nlsteers) [inline acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/utilities/Driver.java:105]: snapshotting the set before iterating would be a safer operation

File: acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/pages/DLCommonPageObject.java [hunk 14]
```diff
@@ -94,19 +93,19 @@ public void navigateToIPVCoreStub() {
...
+            BrowserUtils.clickAndWaitForNavigation(drivingLicenceCRIDev);
         } else if (dlCRITestEnvironment.equalsIgnoreCase("Build")) {
-            drivingLicenceCRIBuild.click();
+            BrowserUtils.clickAndWaitForNavigation(drivingLicenceCRIBuild);
         } else if (dlCRITestEnvironment.equalsIgnoreCase("Staging")) {
-            drivingLicenceCRIStaging.click();
+            BrowserUtils.clickAndWaitForNavigation(drivingLicenceCRIStaging);
         } else if (dlCRITestEnvironment.equalsIgnoreCase("Integration")) {
-            drivingLicenceCRIIntegration.click();
+            BrowserUtils.clickAndWaitForNavigation(drivingLicenceCRIIntegration);
```
Comment (nlsteers) [inline acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/pages/DLCommonPageObject.java:108]: could this not be a switch block instead of a chained if-else?

File: acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/pages/DLCommonPageObject.java [hunk 15]
```diff
@@ -117,6 +116,7 @@ public void navigateToDrivingLicenceCRIOnTestEnv() {
     public void searchForUATUser(String number) {
         selectRow.sendKeys(number);
         searchButton.click();
```
Comment (nlsteers) [inline acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/pages/DLCommonPageObject.java:118]: here would actually be a useful case for a comment explaining why this `.click()` is different to all the others (no staleness wait)

File: acceptance-tests/Dockerfile [hunk 16]
```diff
@@ -1,25 +1,57 @@
...
+COPY "acceptance-tests/public-keys/google-linux-signing-key.pub" "google-linux-signing-key.pub"
+
+# NOTE : Chrome not available for arm64 linux currently and we want to use chrome not chromium
+# 1) Adds Googles gpg key, then sets up the Google Chrome RPM repo, then installs chrome from the repo
+# 2) Import the aws-cli.gpg key, verifies the awscliv2.zip signature and installs the aws-cli
+# 3) Remove the downloaded files
+# 4) Clear the dnf cache to avoid bloating the image any further
+
+RUN dnf install -y --allowerasing --setopt=install_weak_deps=False jq gnupg2 \
+    && gpg --dearmor < google-linux-signing-key.pub > /etc/pki/rpm-gpg/google-chrome.gpg \
```
Comment (nlsteers) [inline acceptance-tests/Dockerfile:27]: RPM should take the key in armoured format, this step is probably not needed

something like this might work

```
&& cp google-linux-signing-key.pub /etc/pki/rpm-gpg/google-chrome \
&& echo -e "[google-chrome]\nname=google-chrome\nbaseurl=http://dl.google.com/linux/chrome/rpm/stable/x86_64\nenabled=1\ngpgcheck=1\ngpgkey=file:///etc/pki/rpm-gpg/google-chrome" > /etc/yum.repos.d/google-chrome.repo \
```

File: acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/pages/UniversalSteps.java [hunk 17]
```diff
@@ -34,7 +34,7 @@ public class UniversalSteps {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(UniversalSteps.class);
 
-    public static final int MAX_WAIT_SEC = 15;
+    public static final int MAX_WAIT_SEC = 60;
```
Comment (nlsteers) [inline acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/pages/UniversalSteps.java:37]: this is a large jump and will significantly lengthen the feedback loop when tests do legitimately fail, it'll also mask any real latency regressions.

i get the need to fix for slow containers in staging but this is a sledgehammer approach

File: .github/workflows/post-merge-deploy-to-dev.yml (hunk 8 above)
Comment (smsgds) [reply .github/workflows/post-merge-deploy-to-dev.yml]: Pr to correct cosign is not merged so this action cannot be used -https://github.com/govuk-one-login/devplatform-upload-action-ecr/pull/24

I have updated our steps to use the latest cosign with the correct way to turn of rekor.

File: .github/workflows/post-merge-deploy-acm-to-dev.yml (hunk 9 above)
Comment (smsgds) [reply .github/workflows/post-merge-deploy-acm-to-dev.yml:33]: We want to pin the version of sam, or we will be back failing randomly as new rules are added as the version floats.

File: acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/utilities/Driver.java (hunk 11 above)
Comment (smsgds) [reply acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/utilities/Driver.java]: You have given me the motivation now to go and clean up this Driver management.

For context our tests used to create and close the driver in every test.
1 - this was very slow
2 - this caused lots focus stealing when running locally

I have removed the close driver after test step now, it was made a No Op (just a print) as we confirmed we had no issues with a single driver.

File: acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/pages/DLCommonPageObject.java (hunk 14 above)
Comment (smsgds) [reply acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/pages/DLCommonPageObject.java]: Some Java11 era code, have updated this.

File: .github/workflows/post-merge-package-for-build.yml
Comment (gds-jbentham) [inline .github/workflows/post-merge-package-for-build.yml:36]: We reuse lines 20 to 40 a lot. Not for this PR but might be worth some debt to make this a composite action

File: acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/pages/DLCommonPageObject.java (hunk 15 above)
Comment (smsgds) [reply acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/pages/DLCommonPageObject.java]: Going to add one to that click.

Its a bit of an edge case but would cover the core stub being slow to redirect us.
Should ensure our url timer is started after we have left the stub page.

File: acceptance-tests/Dockerfile (hunk 16 above)
Comment (smsgds) [reply acceptance-tests/Dockerfile]: Have removed the dearmor line and put this RUN inside a heredoc, as it was getting abit condensed.

File: acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/pages/UniversalSteps.java (hunk 17 above)
Comment (smsgds) [reply acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/pages/UniversalSteps.java]: It balancing act between the cost of upping the size of the cluster enough in dev to meet 15 seconds always, and accepting that failures in the pipelines will cause the test step to take longer. The upside is that the test step should be a more resilient in the envs like staging that have smaller clusters when there is more traffic.

Comment (gds-jbentham) [reply acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/pages/UniversalSteps.java]: If we want to investigate can we get some debt on the board as I would like to revisit this within the context of scaling and test strategy

File: acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/pages/UniversalSteps.java [hunk 18]
```diff
@@ -84,11 +80,11 @@ public static String getJsonPayload(String fileName) {
         }
     }
 
-    private HttpResponse<String> sendHttpRequest(HttpRequest request)
+    private static HttpResponse<String> sendHttpRequest(HttpRequest request)
```
Comment (gds-jbentham) [inline acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/pages/UniversalSteps.java]: Minor gripe  but can we put this below the public methods using it

File: acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/utilities/BrowserUtils.java [hunk 19]
```diff
@@ -347,16 +311,22 @@ public static void verifyElementDisplayed(WebElement element) {
         }
     }
 
+    public static void clickAndWaitForNavigation(WebElement element) {
+        WebElement htmlElement = Driver.get().findElement(By.tagName("html"));
```
Comment (gds-jbentham) [inline acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/utilities/BrowserUtils.java]: This not the easiest read can we rename this "htmlElement" to "pageContainingElement"

File: acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/utilities/Driver.java [hunk 20]
```diff
@@ -11,95 +10,82 @@
...
 
-                case "edge":
-                    if (!System.getProperty("os.name").toLowerCase().contains("windows"))
-                        throw new WebDriverException("Your OS doesn't support Edge");
-                    WebDriverManager.edgedriver().setup();
-                    driverPool.set(new EdgeDriver());
-                    break;
+    private static WebDriver createDriver() {
+        String browser = ConfigurationReader.getBrowser();
+        System.setProperty("webdriver.chrome.logfile", "chromedriver.log");
```
Comment (gds-jbentham) [inline acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/utilities/Driver.java]: Should these be in the chrome case?

File: acceptance-tests/traffic-tests/run-tests.sh [hunk 21]
```diff
@@ -51,13 +49,10 @@ else
...
-
-  sleep 2
-done
-
-cp -r /home/gradle/build/test-results "$REPORT_DIR"
+# The @traffic test suite takes around 6mins to complete
+# It must not run beyond the canary duration,
+# or will be running at the same time as the AC test suite.
+# Leading to the smaller dev F.E containers getting overwhelmed.
+pushd /home/gradle/cri/acceptance-tests
```
Comment (gds-jbentham) [inline acceptance-tests/traffic-tests/run-tests.sh]: Can we create some more debt on this as this works because of the size of the test pack which is brittle

File: acceptance-tests/.dockerignore [hunk 22]
```diff
@@ -1,6 +0,0 @@
-Dockerfile*
```
Comment (gds-jbentham) [inline acceptance-tests/.dockerignore:1]: Why did we remove this file?

Comment (smsgds) [reply acceptance-tests/.dockerignore:1]: AC test are now aligned with Fraud CRI, where the AC test are now part of the top level gradle build (not a standalone project that just lives in the repo) - This is so we can have one libs.toml.yml and avoid issues like duplicate dependabot pr's for the dependancies with separate projects.

So the docker image is built from the root (.dockeringore is now there)

File: lambdas/drivingpermitcheck/src/main/java/uk/gov/di/ipv/cri/drivingpermit/api/domain/DocumentCheckResult.java [hunk 23]
```diff
@@ -8,13 +8,11 @@ public class DocumentCheckResult {
 
     private boolean executedSuccessfully;
     private String result;
-    private String errorMessage;
     private String transactionId;
 
     private CheckDetails checkDetails;
 
     private boolean isValid;
-    private int attemptCount;
```
Comment (gds-jbentham) [inline lambdas/drivingpermitcheck/src/main/java/uk/gov/di/ipv/cri/drivingpermit/api/domain/DocumentCheckResult.java:17]: Was this not used?

File: lib-dva/src/main/java/uk/gov/di/ipv/cri/drivingpermit/library/dva/configuration/DvaCryptographyServiceConfiguration.java (hunk 5 above)
Comment (gds-jbentham) [reply lib-dva/src/main/java/uk/gov/di/ipv/cri/drivingpermit/library/dva/configuration/DvaCryptographyServiceConfiguration.java:66]: We could swap some of these for ACM if we did some smart importing in lower envs but effort invested hinges on the timeline for DVA moving to something better

File: lambdas/drivingpermitcheck/src/main/java/uk/gov/di/ipv/cri/drivingpermit/api/domain/DocumentCheckResult.java (hunk 23 above)
Comment (smsgds) [reply lambdas/drivingpermitcheck/src/main/java/uk/gov/di/ipv/cri/drivingpermit/api/domain/DocumentCheckResult.java:17]: It was never used, attemptCount tracking was placed in the sessionItem where it is used for the same purpose in all CRI's.
https://github.com/govuk-one-login/ipv-cri-dl-api/blob/b92001480bebdc6bcc887bda40af1c170a15478f/lambdas/drivingpermitcheck/src/main/java/uk/gov/di/ipv/cri/drivingpermit/api/handler/DrivingPermitHandler.java#L232


https://github.com/govuk-one-login/ipv-cri-uk-passport-api/blob/041468648e81ac55f82c49de9e6e3e6e28a6a13d/lambdas/checkpassport/src/main/java/uk/gov/di/ipv/cri/passport/checkpassport/handler/CheckPassportHandler.java#L194

File: acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/pages/UniversalSteps.java (hunk 18 above)
Comment (smsgds) [reply acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/pages/UniversalSteps.java]: Have remove this method, there is a duplicate in BrowserUtils. I have also change BrowserUtils.sendHttpRequest so its reusing one client.

File: .github/workflows/post-merge-package-for-build.yml [hunk 24]
```diff
@@ -14,18 +16,16 @@ jobs:
       ENVIRONMENT: build
     permissions:
       id-token: write
-      contents: read
```
Comment (JessWinterborne) [inline .github/workflows/post-merge-package-for-build.yml:17]: I imagine we would still need `contents: read` here.

Comment (smsgds) [reply .github/workflows/post-merge-package-for-build.yml:17]: TY have aligned with dev

Comment (smsgds) [reply .github/workflows/post-merge-package-for-build.yml:17]: I'll recheck these actions, as i have added permissions: {} so there maybe a few places now that its needed

File: acceptance-tests/run-tests.sh [hunk 25]
```diff
@@ -1,7 +1,24 @@
-#!/usr/bin/env bash
```
Comment (JessWinterborne) [inline acceptance-tests/run-tests.sh:1]: Why are we removing the shebang here?

File: .github/workflows/check-pr.yml [hunk 26]
```diff
@@ -7,6 +7,8 @@ jobs:
   pre-commit:
     name: pre-commit
     runs-on: ubuntu-latest
+    permissions:
```
Comment (JessWinterborne) [inline .github/workflows/check-pr.yml:10]: If every job needs contents: read it makes more sense to set this at the top workflow level.

File: acceptance-tests/src/test/resources/features/UKDrivingLicence/DVLA/DVLADrivingLicence.feature [hunk 27]
```diff
@@ -251,7 +236,6 @@ Feature: Driving Licence Test
     And User re-enters postcode as <InvalidPostCode>
     When User clicks on continue
```
Comment (JessWinterborne) [inline acceptance-tests/src/test/resources/features/UKDrivingLicence/DVLA/DVLADrivingLicence.feature]: Should we be waiting for page reload here also?

File: .sdkmanrc [hunk 28]
```diff
@@ -1,5 +1,5 @@
 # Given you are using sdkman (https://sdkman.io/) to install and manage multiple java distributions
 # Enable sdkman_auto_env=true for automatic switching of java version
 # The default location of the configuration file is $HOME/.sdkman/etc/config
-java=17.0.9-amzn
-gradle=8.13
+java=25.0.2-amzn
+gradle=9.4.1
```
Comment (JessWinterborne) [inline .sdkmanrc:5]: Could we update the README to reflect this please? (I think it mentions gradle 8)

File: .gitignore [hunk 29]
```diff
@@ -1,3 +1,5 @@
+.ai
```
Comment (JessWinterborne) [inline .gitignore:2]: What creates .ai ? If its a local tool it might be better off in your global git ignore, but if its from our agreed team tooling then maybe we could add a comment.

**Review by JessWinterborne — COMMENTED** [review]
I haven't made it through all the files (but the majority), submitting review so you can see my comments.

File: .sdkmanrc (hunk 28 above)
Comment (smsgds) [reply .sdkmanrc:5]: Have aligned this with fraud cri so we dont have to keep updating the readme for version bumps

File: acceptance-tests/src/test/resources/features/UKDrivingLicence/DVLA/DVLADrivingLicence.feature (hunk 27 above)
Comment (smsgds) [reply acceptance-tests/src/test/resources/features/UKDrivingLicence/DVLA/DVLADrivingLicence.feature]: Have replaced all these with `user_clicks_on_continue_and_waits_for_page_navigation` to match the underlying method name.

**Review by gds-jbentham — APPROVED** [review]
